File: src/server/handlers/api/v1/graphql/types/note_revision.rs
use chrono::{DateTime, Utc};

pub struct NoteRevision {
    pub id: uuid::Uuid,
    pub summary: Option<String>,
    pub text_for_search: String,
    pub content_type: String,
    pub content: serde_json::Value,
    pub attributes: serde_json::Value,
    pub started_at: Option<DateTime<Utc>>,
    pub written_at: DateTime<Utc>,
    pub inserted_at: DateTime<Utc>,
}

#[async_graphql::Object]
impl NoteRevision {
    async fn id(&self) -> &uuid::Uuid {
        &self.id
    }

    async fn summary(&self) -> Option<&str> {
        self.summary.as_deref()
    }

    async fn text_for_search(&self) -> &str {
        &self.text_for_search
    }

    async fn content_type(&self) -> &str {
        &self.content_type
    }

    async fn content(&self) -> async_graphql::Json<&serde_json::Value> {
        async_graphql::Json(&self.content)
    }

    async fn attributes(&self) -> async_graphql::Json<&serde_json::Value> {
        async_graphql::Json(&self.attributes)
    }

    async fn started_at(&self) -> Option<DateTime<Utc>> {
        self.started_at
    }

    async fn written_at(&self) -> DateTime<Utc> {
        self.written_at
    }

    async fn inserted_at(&self) -> DateTime<Utc> {
        self.inserted_at
    }
}

File: src/server/handlers/api/v1/graphql/types/note_revision.ts
import { builder } from "../builder.ts"

const FORCE_NOT_NULLABLE = { nullable: false } as any

export const NoteRevision = builder.drizzleObject("noteRevisionsTable", {
    variant: "NoteRevision",
    fields: t => ({
        id: t.exposeID("id", { nullable: false }),
        summary: t.exposeString("summary"),
        textForSearch: t.exposeString("textForSearch", { nullable: false }),
        contentType: t.exposeString("contentType", { nullable: false }),
        content: t.expose("content", { type: "JSON", ...FORCE_NOT_NULLABLE }),
        attributes: t.expose("attributes", { type: "JSON", ...FORCE_NOT_NULLABLE }),
        startedAt: t.expose("startedAt", { type: "DateTime" }),
        writtenAt: t.expose("writtenAt", { type: "DateTime", nullable: false }),
        insertedAt: t.expose("insertedAt", { type: "DateTime", nullable: false }),
    }),
})
